components/org.wso2.identity.event.websubhub.publisher/src/main/java/org/wso2/identity/event/websubhub/publisher/service/WebSubHubAdapterServiceImpl.java

```diff
@@ -0,0 +1,329 @@
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleServerException;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.logDiagnosticFailure;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.logDiagnosticSuccess;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.logPublishingEvent;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.parseEventHubResponse;
+
+
+/**
+ * OSGi service for publishing events using web sub hub.
+ */
+public class WebSubHubAdapterServiceImpl implements EventPublisher {
+
+    private static final Log log = LogFactory.getLog(WebSubHubAdapterServiceImpl.class);
+    private String webSubHubBaseUrl = null;
+
+    @Override
+    public void publish(SecurityEventTokenPayload eventPayload, EventContext eventContext)
+            throws WebSubAdapterException {
+
+        makeAsyncAPICall(eventPayload, eventContext,
+                constructHubTopic(eventContext.getEventUri(), eventContext.getTenantDomain()), getWebSubBaseURL());
+        log.debug("Event published successfully to the WebSub Hub.");
+    }
+
+    /**
+     * Register a topic in the WebSub Hub.
+     *
+     * @param eventUri     Event URI.
+     * @param tenantDomain Tenant domain.
+     * @throws WebSubAdapterException If an error occurs while registering the topic.
+     */
+    public void registerTopic(String eventUri, String tenantDomain) throws WebSubAdapterException {
+
+        makeTopicMgtAPICall(constructHubTopic(eventUri, tenantDomain), getWebSubBaseURL(),
+                WebSubHubAdapterConstants.Http.REGISTER);
+        log.debug("WebSub Hub Topic registered successfully for the event: " + eventUri + " in tenant: " +
+                tenantDomain);
+    }
+
+    /**
+     * Deregister a topic in the WebSub Hub.
+     *
+     * @param eventUri     Event URI.
+     * @param tenantDomain Tenant domain.
+     * @throws WebSubAdapterException If an error occurs while deregistering the topic.
+     */
+    public void deregisterTopic(String eventUri, String tenantDomain) throws WebSubAdapterException {
+
+        makeTopicMgtAPICall(constructHubTopic(eventUri, tenantDomain),
+                getWebSubBaseURL(), DEREGISTER);
+    }
+
+    private String getWebSubBaseURL() throws WebSubAdapterException {
+
+        if (StringUtils.isEmpty(webSubHubBaseUrl)) {
+            webSubHubBaseUrl =
+                    WebSubHubAdapterDataHolder.getInstance().getAdapterConfiguration().getWebSubHubBaseUrl();
+
+            // At this point, url shouldn't be null since if adapter is enabled, url is mandatory to configured.
+            // But adding this as a second level verification.
+            if (StringUtils.isEmpty(webSubHubBaseUrl)) {
+                log.warn("WebSubHub Base URL is empty. WebSubHubEventPublisher will not engage.");
```

Comment:
Is this warn log necessary given the exception that is thrown

Reply:
Addressed with https://github.com/wso2-extensions/identity-event-publishers/pull/1/commits/7999427bc5211016a93107fb27f9efd67513d525